### PR TITLE
Make Problems immutable

### PR DESCRIPTION
**Pull request by reinsch82**
Implement #26887

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java [hunk 1]
```diff
@@ -94,4 +94,12 @@ public interface Problem {
...
      */
     Map<String, String> getAdditionalData();
+
+    /**
+     * create a clone builder for this problem.
+     *
+     * @since 8.6
+     */
+    @Incubating
+    ProblemCloneBuilder createCloneBuilder();
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java:104]: `toBuilder()` sounds more concise.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java [hunk 2]
```diff
@@ -94,4 +94,12 @@ public interface Problem {
      * In a map so it can be somewhat structured.
      */
     Map<String, String> getAdditionalData();
+
+    /**
+     * create a clone builder for this problem.
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java:99]:
```suggestion
     * Returns a problem builder with fields initialized with values from this instance.
```

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+package org.gradle.api.problems;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.problems.locations.ProblemLocation;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This is a copy of the ProblemCloneBuilder interface from platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java:26]: 🤔 Can we merge the implementation?

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ */
+@Incubating
+public interface ProblemCloneBuilder extends BuildableProblemBuilder,
+    ProblemBuilderDefiningDocumentation,
+    ProblemBuilderDefiningLocation,
+    ProblemBuilderDefiningLabel,
+    ProblemBuilderDefiningCategory {
+    ProblemCategory getCategory();
+
+    String getLabel();
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java:38]: ❓ 
Why do you need the getters here?

File: platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/PluginIdLocationTransformer.java [hunk 5]
```diff
@@ -36,20 +37,21 @@ public PluginIdLocationTransformer(
...
+        return getExecuteTask(OperationDetails.class)
+            .map(id -> {
                 try {
                     OperationDetails operationDetails = operationListener.getOp(id, OperationDetails.class);
                     Objects.requireNonNull(operationDetails, "operationDetails should not be null");
                     String pluginId = operationDetails.getPluginId();
                     if (pluginId != null) {
-                        problem.getLocations().add(new PluginIdLocation(pluginId));
+                        ProblemCloneBuilder cloneBuilder = problem.createCloneBuilder();
+                        cloneBuilder.getLocations().add(new PluginIdLocation(pluginId));
```
Comment (donat) [inline platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/PluginIdLocationTransformer.java:48]: ❌ 
This is weird. Can we replace it with `cloneBuilder.pluginLocation(pluginId)`? cloneBuilder implements `ProblemBuilderDefiningLocation`, right? If the method does not exist there, I think it should.

File: platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/StackLocationTransformer.java [hunk 6]
```diff
@@ -39,7 +40,9 @@ public Problem transform(Problem problem) {
         ProblemDiagnostics problemDiagnostics = problemStream.forCurrentCaller(problem.getException());
         Location loc = problemDiagnostics.getLocation();
         if (loc != null) {
-            problem.getLocations().add(new FileLocation(loc.getSourceLongDisplayName().getDisplayName(), loc.getLineNumber(), null, null));
+            ProblemCloneBuilder cloneBuilder = problem.createCloneBuilder();
+            cloneBuilder.getLocations().add(new FileLocation(loc.getSourceLongDisplayName().getDisplayName(), loc.getLineNumber(), null, null));
```
Comment (donat) [inline platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/StackLocationTransformer.java:44]: ❌ 
Similar to above.

File: platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/TaskPathLocationTransformer.java [hunk 7]
```diff
@@ -37,18 +38,18 @@ public TaskPathLocationTransformer(
...
-            .ifPresent(id -> {
+        return getExecuteTask(ExecuteTaskBuildOperationDetails.class)
+            .map(id -> {
                 try {
                     ExecuteTaskBuildOperationDetails executeTaskDetails = operationListener.getOp(id, ExecuteTaskBuildOperationDetails.class);
                     Objects.requireNonNull(executeTaskDetails, "executeTaskDetails should not be null");
                     Path taskPath = executeTaskDetails.getTask().getIdentityPath();
-                    problem.getLocations().add(new TaskPathLocation(taskPath));
+                    ProblemCloneBuilder cloneBuilder = problem.createCloneBuilder();
+                    cloneBuilder.getLocations().add(new TaskPathLocation(taskPath));
```
Comment (donat) [inline platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/TaskPathLocationTransformer.java:48]: And here as well.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java (hunk 3 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java:26]: That's just auto generated doc :(.

I'll review the doc again.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java (hunk 4 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemCloneBuilder.java:39]: I thought we should be able do consider the previous values when adding new ones.
But thinking about it, there is always the original problem, we can as for those :)

**Comment by reinsch82** [comment]
@bot-gradle test PT

**Comment by reinsch82** [comment]
continued in 
- #27013

**Pull request closed without merging**